File: Kattis/Oil/Oil.cpp
#include <iostream> 
#include <iterator> 
#include <map>
#include <vector>
#include <algorithm>
using namespace std;
int n,l,r,h,w,x0,_y0,x,y,res=0,run;
int main(){
	map<pair<int, int>, int> lp, rp;
	map<pair<int, int>, int>::iterator itr;
	vector<pair<int, int> > v;
	vector<pair<double, int> > rest;
	cin >> n;
	for (int i=0;i<n;i++){
		cin >> l >> r >> h;
		if (l>r){
			int tmp = r; r = l; l = tmp;
		}
		w = r-l;
		lp[pair<int, int>(l, h)] = w;
		rp[pair<int, int>(r, h)] = w;
		v.push_back(pair<int, int>(l,h));
		v.push_back(pair<int, int>(r,h));
	}
	for (int i=0;i<v.size();i++){
		x = v[i].first; y = v[i].second; rest.clear();
		for (itr = lp.begin(); itr != lp.end(); ++itr) { 
			x0 = itr->first.first; _y0 = itr->first.second; w = itr->second;
			if (_y0 > y) rest.push_back(pair<double, int> (((double) (x0-x))/(_y0-y), -w));
			else if (_y0 < y) rest.push_back(pair<double, int> (((double) (x0-x))/(_y0-y), w));
		}
		for (itr = rp.begin(); itr != rp.end(); ++itr) { 
			x0 = itr->first.first; _y0 = itr->first.second; w = itr->second;
			if (_y0 > y) rest.push_back(pair<double, int> (((double) (x0-x))/(_y0-y), w));
			else if (_y0 < y) rest.push_back(pair<double, int> (((double) (x0-x))/(_y0-y), -w));
		}
		if (!rest.empty()) sort(rest.begin(), rest.end());
		if (lp.find(v[i]) != lp.end()) run = lp[v[i]];
		else run = rp[v[i]];
		res = max(res, run);
		for (int j=0;j<(int) rest.size()-1;j++){
			run -= rest[j].second;
			res = max(res, run);
		}
	}
    cout << res << endl; 
}

File: Kattis/Oil/Oil.py
n = int(raw_input())
lp, rp, res = {}, {}, 0
for i in range(n):
	l, r, h = [int(_) for _ in raw_input().split()]
	l, r = min(l, r), max(l, r)
	w = r-l
	lp[(l, h)] = w
	rp[(r, h)] = w
for x, y in list(lp.keys())+list(rp.keys()):
	rest = []
	for pt, w in lp.items():
		x0, y0 = pt
		if y0>y:
			rest.append(((x0-x)/float(y0-y), -w)) #float may break
		elif y0<y:
			rest.append(((x0-x)/float(y0-y), w))
	for pt, w in rp.items():
		x0, y0 = pt
		if y0>y:
			rest.append(((x0-x)/float(y0-y), w))
		elif y0<y:
			rest.append(((x0-x)/float(y0-y), -w))
	rest.sort()
	run = lp[(x, y)] if (x, y) in lp else rp[(x,y)]
	res = max(res, run)
	for i in range(len(rest)-1):
		run -= rest[i][1]
		res = max(res, run)
print (res)
